devices/milightRF24.coffee

```diff
@@ -0,0 +1,210 @@
+                    self.turnOff(false)
+                  when (Buttons.AllOn or Buttons.Group1On or Buttons.Group2On or Buttons.Group3On or Buttons.Group4On) and data.longPress is true
+                    self.setWhite(false)
+                  when Buttons.ColorFader or Buttons.FaderReleased
+                    self.setColor("#"+self._num2Hex(data.color.r)+self._num2Hex(data.color.g)+self._num2Hex(data.color.b), false)
+                  when Buttons.BrightnessFader or Buttons.FaderReleased
+                    self.setBrightness(data.brightness, false)
+                  
+                return yes
+                
+              return no
+            
+          if br is yes
+            break
+      )
+
+    _num2Hex: (s) ->
+      a = s.toString(16);
+      if (a.length % 2) > 0 
+        a = "0" + a;
+      a;
+      
+    _updateState: (attr) ->
+      state = _.assign @getState(), attr
+      super null, state
+
+    turnOn: (send) ->
+      if typeof send is "undefined"
+        send = true
+        
+      env.logger.debug "Turn on"
+      self = @
+      @_updateState power: true
+      
+      for z in @zones
+        do (z) =>
+          unless z.send is false or send is false
+            self.gateway.turnOn(z.addr, z.zone)
+            if self.mode
+              color = Color(self.color).rgb()
+              self.gateway.setColor(z.addr, z.zone, color.r, color.g, color.b, true)
+            else
+              self.gateway.setWhite(z.addr, z.zone)
+              
+            self.gateway.setBrightness(z.addr, z.zone, self.brightness)
+        
+      Promise.resolve()
+
+    turnOff: (send) ->
+      if typeof send is "undefined"
+        send = true
+        
+      self = @
+      @_updateState power: false
+      for z in @zones
+        do (z) =>
+          unless z.send is false or send is false
+            self.gateway.turnOff(z.addr, z.zone)
+      Promise.resolve()
+
+    setColor: (newColor, send) ->
+      if typeof send is "undefined"
+        send = true
+        
+      self = @
+      color = Color(newColor).rgb()
+      @_updateState
+        mode: @COLOR_MODE
+        color: color
+      
+      for z in @zones
+        do (z) =>
+          unless z.send is false or send is false
+            self.gateway.setColor(z.addr, z.zone, color.r, color.g, color.b, true) if self.power
+      Promise.resolve()
+
+    setWhite: (send) ->
+      if typeof send is "undefined"
+        send = true
+        
+      self = @
+      @_updateState mode: @WHITE_MODE
+      
+      for z in @zones
+        do (z) =>
+          unless z.send is false or send is false
+            self.gateway.setWhite(z.addr, z.zone) if self.power
+      Promise.resolve()
+
+    setBrightness: (newBrightness, send) ->
+      if typeof send is "undefined"
+        send = true
+        
+      self = @
+      @_updateState brightness: newBrightness
+      for z in @zones
+        do (z) =>
```

Comment:
there are a lots of "dos" the the code :) and for this example it is not really necessary, since there is no asynchronous code in the loop and you only properties of the object z.   

I personally would prefer do it like `@zones.forEach (z) ->`. Than you also don't have to thing about if you need a "do" or not. But ok in the end its just a question about style and save lines of code ;-)